GTO basis order (#83)
pyscf/pyscf.github.io: source/user/gto.rst

See :source:`examples/gto/04-input_basis.py` for more examples.

Ordering of basis functions
---------------------------

> GTO basis functions are sorted in the following order: (1) atoms, (2) angular

Reviewer: ```suggestion
GTO basis functions are stored in the following order: (1) atoms, (2) angular
```

> momentum, (3) shells, (4) spherical harmonics. Particularly, basis functions are

Reviewer: ```suggestion
 momentum, (3) shells, (4) spherical harmonics. This means that basis functions are first
```

grouped in terms of the atoms they are assigned to. On each atom, basis

> functions are grouped according to their angular momentum. In each group of
> angular momentum, the ordering index is shell. A shell can be the atomic shell
> or just the group of primitive Gaussian functions with the same exponent. Shells
> are sorted from inner shell to outer shell (or from large exponents to small
> exponents). In each shell, the spherical parts of the GTO basis follow the

Reviewer: ```suggestion
functions are grouped according to their angular momentum.
For each value of the angular momentum, the individual shells
are sorted from inner shells to outer shells, that is, from large exponents to small exponents.
A shell can be a real atomic shell, formed as a linear combination of many Gaussians,
or just a single primitive Gaussian function that may have several angular components. 
In each shell, the spherical parts of the GTO basis follow the
```

> Condon-Shortley convention, with the ordering (and phase) the same to
> https://en.wikipedia.org/wiki/Table_of_spherical_harmonics
> except `p` functions. The p functions are sorted in the order of `px`, `py`,
> `pz` instead of the ordering `py`, `pz`, `px` in the table above.

Reviewer: ```suggestion
Condon-Shortley convention, with the ordering (and phase) given in 
`the Wikipedia table of spherical harmonics <https://en.wikipedia.org/wiki/Table_of_spherical_harmonics>`, 
except for the `p` functions for which the order of `px`, `py`, `pz` is used 
instead of the order`py`, `pz`, `px` used in the table above.
```

0 C       1       s
0 C       2       s
0 C       3       s
0 C       4       s
0 C       2       p                  x
0 C       2       p                  y
0 C       2       p                  z
0 C       3       p                  x
0 C       3       p                  y
0 C       3       p                  z
0 C       4       p                  x
0 C       4       p                  y
0 C       4       p                  z
0 C       3       d                  xy
0 C       3       d                  yz
0 C       3       d                  z^2
0 C       3       d                  xz
0 C       3       d                  x2-y2
0 C       4       d                  xy
0 C       4       d                  yz
0 C       4       d                  z^2
0 C       4       d                  xz
0 C       4       d                  x2-y2
0 C       4       f                  -3
0 C       4       f                  -2
0 C       4       f                  -1
0 C       4       f                  0
0 C       4       f                  1
0 C       4       f                  2
0 C       4       f                  3
========= ======= =================  ===============

> Cartesian GTO functions are sorted slightly different inside the shell.
> For Cartesian GTO functions, the basis functions are generated with the code below::

Reviewer: ```suggestion
The order of Cartesian GTOs is generated by the code below::
```

  for lx in reversed(range(l + 1)):
    for ly in reversed(range(l + 1 - lx)):
      lz = l - lx - ly
      basis = 'x' * lx + 'y' * ly + 'z' * lz

> For example the Cartesian `d` functions has the order `xx`, `xy`, `xz`, `yy`, `yz`, `zz`.

Reviewer: ```suggestion
For example, the Cartesian `d` functions are ordered as `xx`, `xy`, `xz`, `yy`, `yz`, `zz`.
```

> The basis order can be verified with the method :func:`Mole.ao_labels()`.

Reviewer: ```suggestion
 The ordering of the basis functions can be verified with the method :func:`Mole.ao_labels()`.
```
